Source/JavaScriptCore/runtime/JSArrayBufferConstructor.cpp

```diff
@@ -104,8 +102,21 @@ EncodedJSValue JSGenericArrayBufferConstructor<sharingMode>::constructImpl(JSGlo
     // https://tc39.es/proposal-resizablearraybuffer/#sec-allocatesharedarraybuffer
     RefPtr<ArrayBuffer> buffer;
     if (maxByteLength) {
-        if (maxByteLength.value() < length)
+        if (maxByteLength.value() < lengthDouble)
             return throwVMRangeError(globalObject, scope, "ArrayBuffer length exceeds maxByteLength option"_s);
+    }
+
+    JSObject* newTarget = asObject(callFrame->newTarget());
+    Structure* structure = JSC_GET_DERIVED_STRUCTURE(vm, arrayBufferStructureWithSharingMode<sharingMode>, newTarget, callFrame->jsCallee());
+    RETURN_IF_EXCEPTION(scope, { });
+
+    size_t length = 0;
+    if (hasArguments) {
+        length = callFrame->uncheckedArgument(0).toTypedArrayIndex(globalObject, "length"_s, lengthDouble);
```

Comment:
Is there any chance we can avoid adding third parameter to `toTypedArrayIndex`? Maybe we can wrap `lengthDouble` in JSValue and call the method on it? It is very confusing to have an optional parameter that changes the method's operand.

Reply:
Thank you for the review! I too would prefer to avoid adding a new argument if possible.

The problem is that `toNumber` has observable side effects, namely it calls `valueOf`. Also, `toTypedArrayIndex` internally calls `toNumber`.

By calling `toNumber` on `length` for the `length < maxByteLength` check, and then calling `toTypedArrayIndex`, we end up calling `toNumber` twice, resulting in `valueOf` being called twice. This behavior causes tests like the one at https://github.com/WebKit/WebKit/blob/284bb1401647bf26edf133c6543caa7f280e94d2/JSTests/stress/v8-resizablearraybuffer-growablesharedarraybuffer.js#L168-L177 to fail.

To avoid this problem, I've added a new argument to reuse the `double` value already obtained by `toNumber`...

Reply:
Ahh sorry, I misunderstood your comment. I have addressed it.

Reply:
@shvaikalesh friendly ping